### PR TITLE
1836 - Join emps/starters/leavers/vacancies to job role estimates

### PR DESCRIPTION
**Pull request by RoyPriceCS50**
## Description
Trello ticket [1836](https://trello.com/c/DzB2MVl3)

Joined ASCWDS workplace emps and s/l/v to the job role estimates in slv pipeline merge job.
Workplace data had been prepped into 15 published roles, whereas job role estimates was still 37 roles. 
Therefore, I have aggregated the job role estimates dataset before joining workplace data to it.
Updated merge validation to reflect new expected row count.

## Testing
- [x] Unit tests passing
- [x] Successful [Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/statemachines/view/arn%3Aaws%3Astates%3Aeu-west-2%3A856699698263%3AstateMachine%3A1836-join-ascwds-Ind-CQC-SLV?type=standard)
- [x] Outputs checked in Athena

See Trello ticket attachment Excel. 
I checked the estimated filled posts per job role between the job role estimates dataset and the slv merged dataset.
(before and after collapsing job role estimates). Got 0 difference.

I then checked the join by checking "does every job role have some ascwds data attached to it", "How many establishmentid's are in job role metadata vs slv merged" and then spot checked two establishments between slv prepared and slv merged.

All showed 0 difference.

## Checklist (delete if not relevant)
- [x] Unit tests added/amended
- [x] Docstrings added/updated
- [x] Updated CHANGELOG
- [x] Code reviewed by AI
- [x] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/fargate/utils/merge_utils.py [hunk 1]
```diff
@@ -1,7 +1,107 @@
...
+    PublishedJobRoleLabels,
+)
+
+# Roles that exist under the same name in both job-role taxonomies in play here: the
+# 37-category scheme job_role_estimates_lf.main_job_role_clean_labelled uses
+# (EstimatedIndCQCFilledPostsByJobRoleCategoricalValues) and the 15-category published
+# scheme cleaned_ascwds_workplace_lf.job_role_label uses (SLVPrepareCategoricalValues).
+# Every other granular role has no equivalent in the published scheme and must be
+# bucketed into one of the 4 'other_*' groups instead.
+ROLES_SHARED_BY_BOTH_JOB_ROLE_TAXONOMIES = {
```
Comment (MarionJHolloway) [inline projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/fargate/utils/merge_utils.py]: We don't want to hard code this here. There is functionality in the prepare step that calculates this list (or something very similar), so I'd say move that into a utils file with broader scope and call it again here to generate this list.

**Review by MarionJHolloway — COMMENTED** [review]
Looking good for starters, but we decided before to avoid hard-coding columns, so that will need a bit of re-working. I suspect this would be easier to PR if you separate it into 2 - One for the aggregation step and one for the merging step.

File: projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/fargate/utils/merge_utils.py (hunk 1 above)
Comment (RoyPriceCS50) [reply projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/fargate/utils/merge_utils.py]: Do you mean reduce_to_published_roles ?

Comment (MarionJHolloway) [reply projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/fargate/utils/merge_utils.py]: Maybe the specific functionality went in the end. But you still have SYNTHETIC_JOB_ROLE_LABELS and published_job_role_labels, so you can work out the list here from them without having to hard code them

File: projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/tests/fargate/utils/test_merge_utils.py [hunk 2]
```diff
@@ -1,16 +1,54 @@
...
-    def test_convert_ascwds_job_role_columns_to_rows(self):
-        pass
+class TestRolesSharedByBothJobRoleTaxonomies:
+    def test_contains_every_role_common_to_both_taxonomies_and_nothing_else(self):
+        granular_roles = set(CatColType.JobRoleEnumType.categories)
+        published_roles = set(CatColType.PublishedJobRoleLabelEnumType.categories)
 
+        for role in granular_roles | published_roles:
+            is_in_both = role in granular_roles and role in published_roles
+            assert (role in job.ROLES_SHARED_BY_BOTH_JOB_ROLE_TAXONOMIES) == is_in_both
```
Comment (RoyPriceCS50) [inline projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/tests/fargate/utils/test_merge_utils.py:23]: Add test to check that len job.ROLES_SHARED_BY_BOTH_JOB_ROLE_TAXONOMIES < 
CatColType.JobRoleEnumType.categories and < CatColType.PublishedJobRoleLabelEnumType.categories

File: projects/_07_workforce_characteristics/unittest_data/polars_slv_test_data.py [hunk 3]
```diff
@@ -380,3 +386,213 @@ class TestPrepareUtilsData:
...
+                IndCQC.id_per_locationid_import_date: [8],
+                SLVCols.job_role_label: [PublishedJobRoleLabels.other_managers],
+                IndCQC.location_id: ["loc8"],
+                IndCQC.cqc_location_import_date: [date(2024, 2, 1)],
+                IndCQC.primary_service_type: [PrimaryServiceType.non_residential],
+                IndCQC.main_job_group_labelled: [JobGroupLabels.managers],
+                METRIC: [None],
+            },
+        ),
+    ]
```
Comment (RoyPriceCS50) [inline projects/_07_workforce_characteristics/unittest_data/polars_slv_test_data.py:670]: Add a test case for two rows with different job roles that end up in different "published job role labels". Do care worker and admin at the same time.

Add test case for different id per location import date integers.

File: projects/_07_workforce_characteristics/unittest_data/polars_slv_test_data.py [hunk 4]
```diff
@@ -380,3 +386,213 @@ class TestPrepareUtilsData:
...
+            id="unpublished_direct_care_role_buckets_into_other_direct_care",
+            input_data={
+                IndCQC.id_per_locationid_import_date: [3],
+                IndCQC.location_id: ["loc3"],
+                IndCQC.cqc_location_import_date: [date(2024, 1, 1)],
+                IndCQC.primary_service_type: [
+                    PrimaryServiceType.care_home_with_nursing
+                ],
+                IndCQC.main_job_role_clean_labelled: [
+                    MainJobRoleLabels.employment_support
```
Comment (RoyPriceCS50) [inline projects/_07_workforce_characteristics/unittest_data/polars_slv_test_data.py]: Pick one that is defo staying as direct care for all time.

File: projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/tests/fargate/test_validate_01_merge.py [hunk 5]
```diff
@@ -1,37 +1,70 @@
...
-        self.source_df = pl.DataFrame(source_rows, source_schema, orient="row")
-        self.compare_df = self.source_df.select([IndCqcColumns.location_id])
+
+class TestCalculateExpectedRowCount:
+    @pytest.mark.parametrize(
+        "location_import_date_ids, expected_row_count",
+        [
+            pytest.param(
+                [1, 1], PUBLISHED_ROLE_COUNT, id="single_group_multiple_granular_roles"
+            ),
```
Comment (RoyPriceCS50) [inline projects/_07_workforce_characteristics/_01_starters_leavers_vacancies/tests/fargate/test_validate_01_merge.py]: Refactor this for clarity.

Put keyword args next to the data.
